docs/RFCS/20181227_follower_reads_implementation.md

```diff
@@ -0,0 +1,298 @@
+on lease holders. This document proposes mechanisms to expose follower reads
+through new SQL syntax as well as tweaks to make the SQL physical planner and kv
+DistSender aware that follower reads are possible. Given the intention to make
+follower reads an enterprise feature, some of the complexity in this proposal
+stems from the need to inject behavior from CCL code.
+
+# Motivation
+
+Given that cockroachdb stores multiple replicas of data, a client might expect
+that it be able to serve reads from any of those replicas. In order to provide
+its high level of consistency and isolation, cockroachdb currently requires that
+all reads for a range go to the current lease holder. For reads against data
+written sufficiently far in the past, consistency and isolation morally should
+not be a concern as no concurrent writes should be possible.  There are many
+queries which do not require a completely up-to-date view of the database such
+as analytical workloads for report generation.  Enabling reads to be performed
+within the same data center can greatly increase throughput for large reads and
+greatly reduce latency for small ones. Increasing performance and lowering cost
+to run large analytical queries is valuable, especially for geo-distributed
+deployments.  Providing a convenient mechanism to request local rather than
+lease-holder historical reads will be a compelling enterprise feature.
+
+# Guide-level explanation
+
+The Follower Reads RFC lays out the mechanisms to understand closed timestamps
+and presents the rules which the store follows to determine whether a replica
+can serve a read. This document deals with how clients can make use of this
+behavior offered by the storage layer.  The work described in the Follower Reads
+RFC provides the mechanism which enables a replica to determine if a query
+timestamp is adequately old to allow for a follower read through the use of its
+`Closed Timestamp` mechanism. That work has already enabled replicas to fulfil
+read requests in the past. In order to expose this functionality to clients all
+that needs to be done is to convince the SQL physical planner to direct
+historical reads to local nodes and to coax the `DistSender` to send requests to
+followers when appropriate.
+
+The `Closed Timestamp` is tracked on a per-range basis which attempts to lag
+behind "real" time by some target duration controlled by the cluster setting
+`kv.closed_timestamp.target_duration`. As of writing this value defaults to 30
+seconds but could likely be lowered to 5-10 seconds (at some threshold it may
+potentially interfere with on-going transactions). This proposal seeks to
+achieve follower reads by employing stateless approximation of when a follower
+read is possible by assuming that a read may be directed to a follower if it
+occurs at some multiple of the target duration which is controlled by a new 
+cluster setting `kv.follower_reads.target_multiple`. While this may ultimately 
```

Comment:
Have you considered consulting the gateway node's closed timestamp subsystem for information about the remote node? For example, if you're on n1 and you're deciding whether you can serve a follower read from a replica on n2, you can look at the closed timestamp of n2 to get a good idea. I think this is superior to making this purely timing-based, in the sense that it has protection against "slow nodes" built in. With a purely timing-based approach, it's more likely that you'll have to add state to remember when a node rejected follower reads that were thought possible.

Reply:
Sorry, I mean that if you have a range that's on n1, n2, n3 and the leader is, say, n1, then you consider n2 and n3 for serving a follower read for timestamps that have closed out at n1 (perhaps with some padding), the idea being that all that's missing for n2 and n3 to serve the follower read is their catching up on the Raft log. There are still plenty of ways to get this wrong (a follower could be waiting for a snapshot, etc) but overall it offers a little more protection than going blindly by target duration.

Reply:
Interestingly the way I started writing this RFC was with the stateless approach as the rejected alternative and a more stateful approach as the proposal. After some talking to Nathan I was encouraged to flip it, if only because the proposed approach here is simpler to implement and can be nuanced later. 

A problem as far as I can tell with relying on node level closed timestamp arises when there's a range that's on n2, n3, and n4 and the gateway node receiving the request is on n1. n1 won't have information on the state of the closed ts for the range in question and thus still falls back to having to guess. Using the closed timestamp tracking on the node level could enable the recent timestamp to be less conservative.

I'm curious how important it is that this feature seek to be as up to date as possible. I wonder if clients opting in to this functionality will make their decision based on whether the data 10s vs 60s stale.

> the idea being that all that's missing for n2 and n3 to serve the follower read is their catching up on the Raft log

When we hit this situation, the receiving replica won't know that the request had some reason to believe that a follower read was safe so this will still result in a NotLeaseHolderError. Are you suggesting that we should provide some mechanism to mark a request as expecting to be safe and then wait at the follower for it to catch up?